Add IR RAW2 Reading / Transmitting.

jazzgil opened the pull request:
Greetings,
 
I've added support for reading and printing out RAW IR codes. It uses a format that is short enough to be sent as a command, but also allows easy inspections and modifications after the code is constructed. I called it 'RAW2'. It prints out a line to the log (Info level) similar to the following:

IRSEND,RAW2,191C1E1.30161H1515171615151,38,267,134

The encoding consists of a B32-Hex digit (or '.' and 2 digits) for each timing value, appended by the frequency, and denominators for Pulse and Blank.

I've also added the decoding and handling of the command to _P035_IRTX.

Cheers,
Gilad

jazzgil committed:
Add IR RAW2 Reading / Transmitting.

jazzgil committed:
Add IR RAW2 Reading / Transmitting.

TD-er commented:
The error reported by Travis:

```
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino: In function 'void displayRawToReadableB32Hex()':
/home/travis/build/letscontrolit/ESPEasy/src/_P016_IR.ino:286:25: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
if (avg == -1) continue;
^
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino: In function 'boolean Plugin_035(byte, EventStruct*, String&)':
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino:204:53: error: operation on 'i' may be undefined [-Werror=sequence-point]
((p++ & 1)? IrBLen : IrPLen);
^
cc1plus: some warnings being treated as errors
*** [.pioenvs/normal_IR_ESP8266_4096/src/ESPEasy.ino.cpp.o] Error 1
========================== [ERROR] Took 38.88 seconds ==========================
```

It is a warning, reported as an error, but I agree on this one, since it is undefined behavior.

Also in P035, you check for `i < IrRaw.length()` , but you may increase up-to 3 positions in that loop.
So you may want to add another check before reading the last 2.
Also the length of IrRaw is const in that check, so assign the length to a `const size_t` and use that value to check.

jazzgil committed:
Add IR RAW2 Reading / Transmitting - fixed ambiguity.

TD-er commented on src/_P035_IRTX.ino:
```diff
@@ -198,10 +198,15 @@ boolean Plugin_035(byte function, struct EventStruct *event, String& string)
+                   if (c != '.')
+                       buf[idx++] = from_32hex(c) * irLen;
+                   else {
+                       buf[idx++] = (from_32hex(IrRaw[i+1]) * 32 + from_32hex(IrRaw[i+2])) * irLen;
```
You still don't check `(i+2) < IrRaw.length()` 
So reading may get out of range and that may result in a crash.

jazzgil replied:
Submitted a fixed for that as well. Thanks!

jazzgil committed:
Add IR RAW2 Reading / Transmitting - added a safety check.

jazzgil commented:
You are right of course. I've fixed the ambiguity, and added the safety check you've suggested. Do I need to issue another PR or is it enough that I've submitted the changes to my fork?

TD-er commented:
Submitted changes will be included in the pull request, so no need for a new PR.

jazzgil commented:
Thanks. What if I have other fixes to P35, unrelated to the RAW encoding. How do I submit those changes without triggering this PR? - sorry I'm a newb to github forking...

jazzgil commented:
Got it - create a new branch.

TD-er commented:
What is it about, those changes?
If it isn't too much of a difference, you may also append to this PR.

jazzgil commented:
It's about adding back the support for specifying bits, as an option. As it is now, ESPEasy dropped the ability to specify number of bits, i.e. calling the IRremoteESP8266 library without providing the bits parameter. This results in requesting the library to supply the default number of bits. This won't do when you need to supply a different number. 

Specifically I needed that for Denon codes. But I guess this applies to all vendors.

jazzgil committed:
Improved IR RAW2 format to support RLE.

jazzgil committed:
Fixed warnings...

jazzgil committed:
Fixed warnings...

jazzgil commented:
For the record, the new RAW2 format is as follows: 
1. `IRSEND,RAW2,<code>,<frequency in KHz>,<pulse multiplier>,<blank multiplier>`
2. Every pulse or blank timing is converted into a multiplier to be multiplied by the corresponding pulse or blank multiplier.
3. The multiplier is then encoded in the code as a single B32 Hex digit ([0-9A-V](https://en.wikipedia.org/wiki/Base32#base32hex)).
4. If the multiplier is greater than 31, it is encoded as two B32 digits with a '^' sign prepended.
5. If 2 multipliers repeat, for example: `454545`, they are replaced by '*' and a single times digit, and then the two multipliers. For eaxmple: `*345`.

Hope this helps anyone...
Gilad

TD-er commented:
It is 4 parameters, so you could also create an event for it.
Then it can be used in rules.

jazzgil commented:
Is is possible to pass a string to an event? 
Can you give an example for an event with parameters?

jimmys01 commented:
if this works it is a better solution that the original RAW solution.  frequency in KHz is always 38khz as far as i know since the decoding in the library is done in this frequency. Also in my opinion if you are going to bring back the need to declare the bits then you also hahe to bring back the repeats back. I suggest that for the sake of simplicity to be left like it is now. If someone needs this functionality can allways sent the RAW or RAW2 codes instead

jazzgil commented:
I agree about the 38Khz - it does seem redundant.
I believe the bits and repeat params should be optional.

jimmys01 commented:
Is it possible to make the encoding instead of Base 32, a Base 64 string in order to save even more space?
Also event generation would be a very nice, we can make the ESP follow commands from an IR remote
I will try to test the RAW2 functionality tommorow, ecpecially how it handles HVAC 99bit codes etc

jazzgil commented:
Base 64 wouldn't save that much, as RAW2 tries to keep readability, and thus keeps each timing value in a separate digit. And digits are mostly low value digits (if the algorithm successfully found a high common denominator).

A 99bit code would mean 198 timing values, which would mean there's no chance to encode that in a RAW2 string that would fit under 100 chars - the buffer limit currently set - unless there's a lot of repeated bits.

I have an idea for a more compressed format, that can utilize the base 64 format, and perhaps be suitable for such long codes.

jazzgil committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into mega

jimmys01 commented:
It works and it is certainly a step in the right direction to be able to capture RAW codes directly from the ESP that are easily transportable via MQTT. 
For the NEC Code : 10EF708F (32 bits)I have captured those strings (88 to 110 chars long)
```
IRSEND,RAW2,^5D8*3B1A3A1B1B1A1B3B3A3A1*4B3B1*3B3B1A1B1B1A3A1B1B1*4A39^29^5C4B,38,52,550
IRSEND,RAW2,^5K^2MCBBBBBB^10CABCBABBB^10C^10B^10BB*4B^11BB*3B^10*3BBD9B^10BBBBCB*2B^10C^10B^10B^O5^5K^1BB,38,50,52
IRSEND,RAW2,^5D^17B5A5B5BEB5A5B5B5*3AFA5BE*3BFB5BFAFCEB5A5A5B5BFB5A5B5*3BFAEB^B1^5CKB,38,52,114
```

And for just 50 bits from my HVAC the string count goes to 90-120 characters 

```
IRSEND,RAW2,DB*3C3C7C7D3C3C3D^17CBC3C3C7C3B3C7C7B7D^17CBD7C7C7D7*3C3D3C^17CBC7D3*4C3C7C3C^17CBC3C7B3C3C3D3C7C3C,38,79,975
1068802: EVENT: IRRX#IRRX=2299527906
1072321: IR: Encoding : UNKNOWN^Code : FF245C58 (50 bits)^
1072414: IRSEND,RAW2,ABA393A3A797A393A3A^169B93A3A7A393*3A7A^16ABA7A79797A393A3A39^16AB97*493A3A793A^16AB93A7A383*4A3A,38,99,996
1072417: EVENT: IRRX#IRRX=4280573016
1074121: IR: Encoding : UNKNOWN^Code : 891002E2 (50 bits)^
1074228: IRSEND,RAW2,1B*3131717*3131^161B1313171313*3171^161B*417*4131^161B17*51317131^161B1317*41317131,38,949,999
1074231: EVENT: IRRX#IRRX=2299527906
1077121: IR: Encoding : UNKNOWN^Code : FF245C58 (50 bits)^
1077204: IRSEND,RAW2,ABC3A3A3A7A7*3A3B^16ABB3A3A7A3A3*3A7A^16ABA7A7B7A7*4A3A^16ABA7*5A3A7A3A^16ABA3A7*6A3A,38,92,978
1077206: EVENT: IRRX#IRRX=4280573016
```

With the previous RAW method the strings where much more compact but you have to resort to excel sheets in order to calculate them.

A workaround is to maximize the current 100 chars limit... Any ideas?

jimmys01 commented:
You give to much fidelity to the RAW code in my opinion, i believe that just finding the minimum Pulse and Blank periods and dividing them by 2 is enough fidelity or even perhaps plenty.

jazzgil commented:
I tried to increase the tolerance (which is currently 7%), and results were not effective. Maybe it is just my satellite device, and other devices are more tolerant.  Tolerance % is calculated as the delta between any original timing, and the result after encoding and decoding.

jimmys01 commented:
I tried another bit obscure remote (the kind that RAW codes are very useful for) and it cannot find a common denominator for any of the keys.

jazzgil commented:
Check out my post in the [forum](https://www.letscontrolit.com/forum/viewtopic.php?f=6&t=5853&sid=a858775c8f4b4e5b50bb944852e3ab25) - I've written about needing several tries some times. It doesn't matter which key - with the IR codes of some remotes the timings are too fluid, and unrelated. If the found common denominator is too small, the result code would be too long, and thus abandoned. But still after many tries I've always succeeded.

jimmys01 commented:
Tried it multiple times per button. Over 10 buttons... No success with this particulat remote. I will post later this day the RAW timings of it.

jazzgil commented:
Interesting. Note that you can see the original raw timings if you set the log level to Debug.

Perhaps we should add a parameter in the device settings for Tolerance %.

jimmys01 commented:
the common denominator approach gives very good signal fidelity but is overcomplicated in my opinion. Please check the method that RAW is calculated [here](https://docs.google.com/spreadsheets/d/16icZJWFbHq2iIxuEYxHzKWoP4ZEtra_iegIqDCMVvG4/edit?usp=sharing) and you will see that the min pulse blank / 2 works fine. The Biggest common denominator method give to small pulse and blank periods and thus the encoding string gets that big

jazzgil commented:
Which code did you have trouble with? Is it the one that is set on the sheet? The calculation in that sheet causes the result to be almost 100% off the real raw numbers - look at the 3rd pulse - the result encoding turns the 530 pulse time into 266. If that works for your device it has a very high tolerance. Can't say the same for some of my devices.

jazzgil committed:
Merge branch 'mega' into add-raw-ir-reading

# Conflicts:
#	src/_P016_IR.ino
#	src/_P035_IRTX.ino

jazzgil committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into add-raw-ir-reading

jimmys01 commented:
How about you round the least significant digit of each pule - blank. This is anyway within the erros of the receiver and the timing of the ESP and you guarandee that at least 1 common denominator will be found (10). 
With the RAW method in that already exists I had great success so far.

jazzgil commented:
Its even easier than that - simply decrease the minimum viable denominator (currently 50) to 10, in 
`while (--cd >= 50)`

But that in some cases will create large encodes. But I agree its better than nothing..

jazzgil commented:
From the testings I just did, I dare say that if a denominator above ~50 is not found, you have a faulty raw IR reading. Even if you transmit the exact timings read, the device will not recognize them. At least that's true for my tv streamer.  That's for the "No proper divisor found" error. The "Raw code too long" error is something else...

jazzgil committed:
Increased buffer size. Added constants.

jazzgil commented:
TD-er - Not sure what's the process here - do I need to do anything else for this PR to get merged into the main branch?

TD-er commented:
Has this been tested by others too? (not only the author of this PR)

TD-er committed:
Merge branch 'mega' into add-raw-ir-reading

jimmys01 commented:
@jazzgil @TD-er 

>Perhaps we should add a parameter in the device settings for Tolerance %. 

Thats actually a very good idea.  
`#define PCT_TOLERANCE       7u`  can this be made to be set from the GUI?

TD-er commented:
Not really read the entire thread.
Can you tell me where such a parameter is needed?
Given that it's about IR, it sounds like it should be part of the RX plugin?
